59.md

```diff
@@ -0,0 +1,55 @@
+NIP-59
+======
+
+Gift Wrap
+-------------------------------
+
+`draft` `optional` `author:kieran`
+
+This NIP defines a simple method to protect event metadata leakage by using randomly generated keys.
+
+A gift wrap event is a `kind 1059` event which wraps another event with a single use key.
+
+This event kind can be used to wrap DM's or other events which are considered private.
+
+This is similar in concept to Onion Routing.
```

Comment:
if the sender of the event would like to decrypt it at a later date, they can use their "tweaked" private key, rather than a random key when producing the event.   the "tweak value" can be located in an optional implementation dependent tag, or it can simply use the "created_at" field.  tweaking secp256k1 private keys is a well-known mechanism, and should be considered if a client wants to be able to read these events later from a message store, for example.

Reply:
Any change in how key material is handled needs to be added to NIP-07 or else it's not possible for me to use, this is why ephemeral keys work for me because i create the secret material.

Reply:
agreed.  that would have to happen first!

Reply:
What do you think of adding a nonce key with a random string to the wrapped event JSON (not as a tag, but as an extra field)?

So for the DM example, you could send the gift wrapped DM with 2 1059 events from different random keys and 2 nonces, one to the recipient and another one to yourself (so you can keep chat history). With 2 nonces, the 2 gift wrap `content`s won't match, so no one can tell it is the same message.

This way you don't need to keep track of the wrapping key for decrypting chat history.

The down side is the created_ats will be close to each other and your copy won't have an `e` tag with the recipient so you cant fetch just the messages sent to a specific user. =(

Reply:
@staab nice so the sender should encrypt the receiver pubkey using the shared secret

But from NIP-04, is `secp256k1.getSharedSecret(SENDER_privkey, '02' + RECEIVER_pubkey).slice(1, 33)` the same value as `secp256k1.getSharedSecret(RECEIVER_privkey, '02' + SENDER_pubkey).slice(1, 33)`? So the receiver can use it to also encrypt his own pubkey and query with `{ #p: [encrypted_receiver_pubkey] }`?

Reply:
No idea, encrypted tags would rely on signing with a shared secret. This PR uses ephemeral keys, but it would be easy to convert it to using a shared secret (since one is necessary for decryption anyway). Encrypted tags could then be encrypted symmetrically.